go/kbfs/libkbfs/subscription_manager.go

```diff
@@ -0,0 +1,366 @@
+	}, limit)
+	sm.nonPathSubscriptionIDToTopic[sid] = topic
+	subscriptionIDSetter()
+	return nil
+}
+
+func (sm *subscriptionManager) unsubscribePath(
+	ctx context.Context, subscriptionID SubscriptionID) {
+	sm.lock.Lock()
+	defer sm.lock.Unlock()
+	ref, ok := sm.pathSubscriptionIDToRef[subscriptionID]
+	if !ok {
+		return
+	}
+	delete(sm.pathSubscriptionIDToRef, subscriptionID)
+	if (sm.pathSubscriptions[ref]) == nil {
+		return
+	}
+	if notifier, ok := sm.pathSubscriptions[ref][subscriptionID]; ok {
+		notifier.shutdown()
+		delete(sm.pathSubscriptions[ref], subscriptionID)
+	}
+	if len(sm.pathSubscriptions[ref]) == 0 {
+		sm.unregisterForChangesLocked(ref.folderBranch)
+		delete(sm.pathSubscriptions, ref)
+	}
+	delete(sm.subscriptionIDs, subscriptionID)
+}
+func (sm *subscriptionManager) unsubscribeNonPath(
+	ctx context.Context, subscriptionID SubscriptionID) {
+	sm.lock.Lock()
+	defer sm.lock.Unlock()
+	topic, ok := sm.nonPathSubscriptionIDToTopic[subscriptionID]
+	if !ok {
+		return
+	}
+	delete(sm.nonPathSubscriptionIDToTopic, subscriptionID)
+	if sm.nonPathSubscriptions[topic] == nil {
+		return
+	}
+	if notifier, ok := sm.nonPathSubscriptions[topic][subscriptionID]; ok {
+		notifier.shutdown()
+		delete(sm.nonPathSubscriptions[topic], subscriptionID)
+	}
+	delete(sm.subscriptionIDs, subscriptionID)
+}
+
+func (sm *subscriptionManager) nodeChangeLocked(node Node) {
+	path, ok := node.GetPathPlaintextSansTlf()
+	if !ok {
+		return
+	}
+	ref := pathSubscriptionRef{
+		folderBranch: node.GetFolderBranch(),
+		path:         cleanInTlfPath(path),
+	}
+	if sm.pathSubscriptions[ref] == nil {
+		return
+	}
+	for _, notifier := range sm.pathSubscriptions[ref] {
+		notifier.notify()
+	}
+}
+
+// SubscribePath implements the Subscriber interface.
+func (s subscriber) SubscribePath(ctx context.Context, sid SubscriptionID,
+	path string, topic keybase1.PathSubscriptionTopic,
+	deduplicateInterval *time.Duration) error {
+	return s.sm.subscribePath(ctx,
+		sid, path, topic, deduplicateInterval, s.notifier)
+}
+
+// SubscribeNonPath implements the Subscriber interface.
+func (s subscriber) SubscribeNonPath(ctx context.Context, sid SubscriptionID,
+	topic keybase1.SubscriptionTopic,
+	deduplicateInterval *time.Duration) error {
+	return s.sm.subscribeNonPath(ctx,
+		sid, topic, deduplicateInterval, s.notifier)
+}
+
+// Unsubscribe implements the Subscriber interface.
+func (s subscriber) Unsubscribe(ctx context.Context, sid SubscriptionID) {
+	s.sm.unsubscribePath(ctx, sid)
+	s.sm.unsubscribeNonPath(ctx, sid)
+}
+
+var _ SubscriptionManagerPublisher = (*subscriptionManager)(nil)
+
+// FavoritesChanged implements the SubscriptionManagerPublisher interface.
+func (sm *subscriptionManager) FavoritesChanged() {
+	sm.lock.RLock()
+	defer sm.lock.RUnlock()
+	if sm.nonPathSubscriptions[keybase1.SubscriptionTopic_FAVORITES] == nil {
+		return
+	}
+	for _, notifier := range sm.nonPathSubscriptions[keybase1.SubscriptionTopic_FAVORITES] {
+		notifier.notify()
```

Comment:
If the new "blocking" mode isn't removed and is used by something, this could block while holding the lock, which seems bad.  I'd argue just to remove "blocking" mode and buffer the channel.